common/src/main/java/com/microsoft/identity/common/crypto/KeyStoreBackedSecretKeyProvider.kt

```diff
@@ -0,0 +1,486 @@
+         * Indicate that token item is encrypted with the key loaded in this class.
+         */
+        const val KEY_TYPE_IDENTIFIER: String = "A001"
+
+        @VisibleForTesting
+        const val KEY_FILE_SIZE: Int = 1024
+
+        /**
+         * SecretKey cache. Maps wrapped secret key file path to the SecretKey.
+         */
+        private val sKeyCacheMap: ConcurrentMap<String, SecretKey> = ConcurrentHashMap()
+    }
+
+    override val keyTypeIdentifier = KEY_TYPE_IDENTIFIER
+    override val cipherTransformation = AES_CBC_PKCS5_PADDING_TRANSFORMATION
+
+    /**
+     * File where the wrapped secret key is stored.
+     */
+    private val keyFile =
+        File(context.getDir(context.packageName, Context.MODE_PRIVATE), filePath)
+
+    @get:VisibleForTesting
+    val keyFromCache: SecretKey?
+        get() {
+            clearCachedKeyIfCantLoadOrFileDoesNotExist()
+            return sKeyCacheMap[filePath]
+        }
+
+    @VisibleForTesting
+    fun clearKeyFromCache() {
+        sKeyCacheMap.remove(filePath)
+    }
+
+    /**
+     * Wipe all the data associated from this key.
+     */
+    @VisibleForTesting
+    @Throws(ClientException::class)
+    fun deleteSecretKeyFromStorage() {
+        AndroidKeyStoreUtil.deleteKey(alias)
+        FileUtil.deleteFile(keyFile)
+        sKeyCacheMap.remove(filePath)
+    }
+
+    private fun clearCachedKeyIfCantLoadOrFileDoesNotExist() {
+        val shouldClearCache = !KeyStoreBackedSecretKeyProviderFactory.skipKeyInvalidationCheck &&
+                (!AndroidKeyStoreUtil.canLoadKey(alias) || !keyFile.exists())
+        if (shouldClearCache) {
+            sKeyCacheMap.remove(filePath)
+        }
+    }
+
+    /**
+     * Returns the secret key. If the key is already cached, it returns the cached key.
+     * If the key is not cached, it tries to read the key from storage.
+     * If the key does not exist in storage, it generates a new secret key and caches it.
+     *
+     * @return SecretKey
+     * @throws ClientException if there is an error reading or generating the key
+     */
+    @get:Throws(ClientException::class)
+    @get:Synchronized
+    override val key: SecretKey
+        get() {
+            val methodTag = "$TAG:getKey"
+
+            keyFromCache?.let {
+                Logger.info(
+                    methodTag,
+                    "Key is already cached, returning cached key with thumbprint: " +
+                            KeyUtil.getKeyThumbPrint(it)
+                )
+                return it
+            }
+
+            readSecretKeyFromStorage()?.let {
+                sKeyCacheMap[filePath] = it
+                Logger.info(
+                    methodTag,
+                    "Key loaded from storage and cached with thumbprint: " +
+                            KeyUtil.getKeyThumbPrint(it)
+                )
+                return it
+            }
+
+            val newKey = generateNewSecretKey()
+            sKeyCacheMap[filePath] = newKey
```

Comment:
>sKeyCacheMap[filePath] = newKey [](http://example.com/codeflow?start=12&length=31)

shall we move this generateNewSecretKey. It writes to file, can as well write to in-memory cache.

Reply:
I put it outside on purpose, to explicitly do the cache operations here in getKey